### ISSUE TITLE
keyof and Lookup types do not infer correctly to generic types

### ISSUE THREAD

**Issue by benbayard** [Microsoft/TypeScript#13201]
**TypeScript Version:**  2.1.4

**Code**

```ts
export class Genericizer<T> {
  constructor(public value: T) {}
}

export interface IFoo {
  fooString: string;
  fooNumber: number;
}

export class Foo {
  fooString: Genericizer<string> = new Genericizer('yay');
  fooNumber: Genericizer<number> = new Genericizer(4);
  fooDontTouch: Genericizer<boolean> = new Genericizer(true);

  updateFoo<K extends keyof IFoo>(key: K, value: IFoo[K]) {
    this[key] = new Genericizer(value);
  }
}

const a = new Foo();
a.updateFoo('fooString', 'yahoo'); // referred to as 1
a.updateFoo('fooNumber', 14); 
a.updateFoo('fooDontTouch', false);
```

**Expected behavior:**
This code should compile and be typesafe. On top of that (1) should not throw an error.

**Actual behavior:**
We receive two compilation errors. The second one is expected and valid. 

```bash
ERROR in index.ts
    (16,5): error TS2322: Type 'Genericizer<IFoo[K]>' is not assignable to type 'this[K]'.
    
ERROR in index.ts
    (23,13): error TS2345: Argument of type '"fooDontTouch"' is not assignable to parameter of type '"fooString" | "fooNumber"'.
```
We technically know that all `IFoo[K]` properties on `Foo` _must_ be `Genericizer` types, with the _same_ type as IFoo. However, the compiler does not seem to infer this. 

There are slight work arounds for this, such as having `updateFoo` take a `Genericizer<T>` instead of a `IFoo[K]`. However, that does not work in more complicated flows, such as promises.

**Comment by benbayard** [comment]
@RyanCavanaugh I just wanted to ping on this before it gets lost. I just testing on TS@next and I am getting the same error.

**Comment by weswigham** [comment]
Consider:
```ts
a.updateFoo('updateFoo', 'look at me, mangling my method');
```

The update method itself is what make the assignment unsafe.

**Comment by weswigham** [comment]
The error is, from what I can tell, due to variance (`this` is a type parameter, after all):
```ts
export class Genericizer<T> {
  constructor(public value: T) {}
}

export interface IFoo {
  fooString: string;
  fooNumber: number;
}

export class Foo {
  fooString: Genericizer<string> = new Genericizer('yay');
  fooNumber: Genericizer<number> = new Genericizer(4);
  fooDontTouch: Genericizer<boolean> = new Genericizer(true);

  updateFoo<K extends keyof IFoo>(key: K, value: IFoo[K]) {
    this[key] = new Genericizer(value);
  }
}

const a = new Foo();
a.updateFoo('fooString', 'yahoo');
a.updateFoo('fooNumber', 14); 
a.updateFoo('fooDontTouch', false);

// Why is this a problem? Consider:

export class SubGenericizer<T> extends Genericizer<T> {
    valueAgain: T;
}

export class Bar extends Foo {
  fooString: SubGenericizer<string> = new SubGenericizer('yay');
  fooNumber: SubGenericizer<number> = new SubGenericizer(4);
}

const b = new Bar();
b.updateFoo('fooString', 'yahoo');
b.fooString.valueAgain; // Expected to be present, but it isn't, since the update method just replaced the subgenericizer with a genericizer!
```
